
# Identifying genomic signal of local adaptation in the invasive Queensland fruit fly, _Bactrocera tryoni_

Here we will use the DarTseq data generated by <a href="https://www.nature.com/articles/s41598-020-67397-5" title="Popa‑Báez et al (2020)" >Popa‑Báez et al (2020)</a> from the native and expanded ranges of _B. tryoni_ to understand population genomic underpinnigs of invasion success. We use genome-wide scan to identify highly differentiated genomic variants (SNPs) associated with the invasive status of _B. tryoni_ populations. Additionally, we determined which environmental variables have a key role in shaping allelic trends across the distribution range of _B. tryoni_.

***************

## Table of content
### 1-variant_calling:   
contains scripts for running BWA MEM, SAMtools, BCFtools and PLINK

### 2-population_structure:
contains scripts for PCA, sNMF, StAMPP, pophelper and DivMigrate + input files

### 3-BayPass
contains script for running the C2-statistic analysis in BayPass + input files

### 4-GOterms
contains scripts for running Interproscan and topGO + input files

### 5-environmental_association
contains scripts for processing bioclimatic variables to be used in gradient forest analysis
